[READY] Added workflow for publishing Docker images via CI/CD

maishiroma opened the pull request:
## Summary
This PR adds in a new CI/CD workflow in the `nix-garage` repository where images defined via `nix` in `imgs/` can be automatically built and published to Nebulaworks's Docker Hub page.

## Workflow
To utilize the new job, the following workflow has been established (feedback is welcome!)
1. A contributor creates some new changes/adds a new image in the `img/` folder.
2. The contributor pushes their code up and opens a new PR to the master repository
3. After reviewing the new code, a person from Nebulaworks's `approvers` team can kick off a CI job by commenting in the PR:  `publish` (as of now, any words after `publish` is ignored, but the first word of the comment _must_ be `publish`) 
4. After a few minutes, a bot account (in this case, `nwiauto`) will automatically comment on that same PR, saying the following items:
    - The job was a success, with all of the images that are in the repo being published.
    - The job failed, with specific remarks to why it failed. 

> The PR comments were created using [this GitHub Action](https://github.com/marketplace/actions/comment-action)

## Requirements
To leverage this workflow, the following pre-requirements must be met:
1. Two secrets declared in the repository's settings that correlate to:
    - `nwiauto`'s Docker Hub Password (named: `NWIAUTO_PASSWORD`)
    - `nwiauto`'s GitHub Personal Access Token (named: `NWIAUTO_PAT`)
        - The permissions on the PAT are: 
```
read:discussion, read:org, read:user, repo
```

## Testing
All functionality was tested in a separate [environment](https://github.com/maishiroma/GitGoodActons), where the underline logic was worked through. Refer to [this PR](https://github.com/maishiroma/GitGoodActons/pull/1) for an example of what a PR would look like and this [job](https://github.com/maishiroma/GitGoodActons/runs/755494884?check_suite_focus=true) for an example of a working output.

sarcasticadmin commented on .github/workflows/publish_imgs.yml:
```diff
@@ -0,0 +1,155 @@
+          USER: nwiauto
+        if: steps.verify.outputs.status == 'success'
+        run: |
+          echo "${{ secrets.NWIAUTO_PASSWORD }}" | docker login docker.io --username "$USER" --password-stdin
```
Ah didnt realize that with the GitHub runner and docker. The password be `encoded` at base64 not something thats encrypted although it might appear that way from an initial glance.

I would just say for simplicity we should just leverage the `echo` and `heredoc`. That way we arent relying on any docker components that were implied from the base vm.

maishiroma replied:
Right, my mistake haha. I agree, in my testing, I just confirmed the heredoc function works, so I'll be sticking with that function for logging in.

sarcasticadmin commented on .github/workflows/publish_imgs.yml:
```diff
@@ -0,0 +1,155 @@
+        run: |
+          resp=$(curl -sSf \
+          --url https://api.github.com/orgs/Nebulaworks/teams/approvers/memberships/${{ github.event.comment.actor.login }} \
+          --header 'Authorization: Bearer ${{ secrets.NWIAUTO_PAT }}' \
```
Ah you raise some good points:

I really dont mind either way for posts from `nwiauto `or the `github-actions-bot`. But I will say that if we could leverage the `GITHUB_TOKEN` for all of this actions functionality it will get us out of managing an additional secret which is definitely a ➕.

So why dont you see if you can get endpoint working for the teams and if so well go with the `github-actions-bot` via the `GITHUB_TOKEN`. If it does not work for the teams endpoint then well keep it as you initial had it with the `NWIAUTO_PAT`. Sound good?

maishiroma replied:
Cool! I did some more testing with other API endpoints, and unfortunately, the `GITHUB_TOKEN` does not work in getting those endpoints I mentioned earlier. At this point, I think sticking with the PAT would be best, even though it is another secret to manage.

sarcasticadmin replied:
Sounds good, thanks for giving it a test 💯

maishiroma commented:
New changes:
- Used heredoc to authenticate to `docker.io` (docker hub)
- Removed pre-req step in favor of using docker container to build out images
- Removed login step and combined it with the publish step
- Removed error step for publishing a comment if login failed

Because we are using Docker containers now to build images, I found a way to "persist" the cache of `nix-shell` so that image building will be a lot faster after the initial download + build of the first image.

sarcasticadmin commented on .github/workflows/publish_imgs.yml:
```diff
@@ -0,0 +1,134 @@
+    outputs:
+      allImgs: ${{ steps.publish.outputs.allImgs }}
+
+    # Utilizing nixos/nix docker image v2.3
```
Nice thanks for adding the version comment

sarcasticadmin commented on .github/workflows/publish_imgs.yml:
```diff
@@ -0,0 +1,134 @@
+
+    # Utilizing nixos/nix docker image v2.3
+    container:
+      image: nixos/nix@sha256:af330838e838cedea2355e7ca267280fc9dd68615888f4e20972ec51beb101d8
```
Pinning the SHA 🎖️

maishiroma commented:
@sarcasticadmin Go ahead with the merge! Just squashed all of the commits.
